src/pymorize/pipeline.py

```diff
@@ -56,12 +58,32 @@ def __str__(self):
             r_val.append(f"[{i+1}/{len(self.steps)}] {step.__name__}")
         return "\n".join(r_val)
 
+    def __getstate__(self):
+        """Custom pickling of a Pipeline"""
+        state = self.__dict__.copy()
+        if self._steps_are_prefectized:
+            state["_steps"] = self._raw_steps
+            del state["_raw_steps"]
+            state["_steps_are_prefectized"] = False
+        # It makes no sense to pickle the cluster
+        del state["_cluster"]
+
+        return state
+
+    def __setstate__(self, state):
+        self.__dict__.update(state)
+        if self._workflow_backend == "prefect":
+            self._prefectize_steps()
+        logger.info("Restoring from pickled state!")
+        logger.info("You may want to assign a cluster to this pipeline")
+
```

Comment:
This part was specific to cache `Rule` objects, since `Rule`s have the `Pipeline` attached to them, and the `Pipeline` in turn has a collection of Prefect `Task` objects. Also, it is not sensible to cache the Dask `SLURMCluster` object, since that is unique to each run.

I'm not sure which one of the two was causing it to be impossible to pickle things.

Reply:
Thanks for pointing out the place. Looks like a good solution to me, since the pipeline tasks will be run anyway with their corresponding caches.